File: lib/src/core/linker/compiler.dart
library angular2.src.core.linker.compiler;

import "package:angular2/src/core/linker/view_ref.dart" show HostViewFactoryRef;
import "package:angular2/src/core/di.dart" show Injectable;
import "package:angular2/src/facade/lang.dart" show Type, isBlank, stringify;
import "package:angular2/src/facade/exceptions.dart" show BaseException;
import "package:angular2/src/facade/async.dart" show Future, PromiseWrapper;
import "package:angular2/src/core/reflection/reflection.dart" show reflector;
import "package:angular2/src/core/linker/view.dart" show HostViewFactory;
import "package:angular2/src/core/linker/view_ref.dart"
    show HostViewFactoryRef_;

/**
 * Low-level service for compiling [Component]s into [ProtoViewRef ProtoViews]s, which
 * can later be used to create and render a Component instance.
 *
 * Most applications should instead use higher-level [DynamicComponentLoader] service, which
 * both compiles and instantiates a Component.
 */
abstract class Compiler {
  Future<HostViewFactoryRef> compileInHost(Type componentType);
  clearCache();
}

bool isHostViewFactory(dynamic type) {
  return type is HostViewFactory;
}

@Injectable()
class Compiler_ extends Compiler {
  Future<HostViewFactoryRef_> compileInHost(Type componentType) {
    var metadatas = reflector.annotations(componentType);
    var hostViewFactory =
        metadatas.firstWhere(isHostViewFactory, orElse: () => null);
    if (isBlank(hostViewFactory)) {
      throw new BaseException(
          '''No precompiled component ${ stringify ( componentType )} found''');
    }
    return PromiseWrapper.resolve(new HostViewFactoryRef_(hostViewFactory));
  }

  clearCache() {}
}

File: es6/prod/src/core/linker/compiler.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Injectable } from 'angular2/src/core/di';
import { isBlank, stringify } from 'angular2/src/facade/lang';
import { BaseException } from 'angular2/src/facade/exceptions';
import { PromiseWrapper } from 'angular2/src/facade/async';
import { reflector } from 'angular2/src/core/reflection/reflection';
import { HostViewFactory } from 'angular2/src/core/linker/view';
import { HostViewFactoryRef_ } from 'angular2/src/core/linker/view_ref';
/**
 * Low-level service for compiling {@link Component}s into {@link ProtoViewRef ProtoViews}s, which
 * can later be used to create and render a Component instance.
 *
 * Most applications should instead use higher-level {@link DynamicComponentLoader} service, which
 * both compiles and instantiates a Component.
 */
export class Compiler {
}
function isHostViewFactory(type) {
    return type instanceof HostViewFactory;
}
export let Compiler_ = class extends Compiler {
    compileInHost(componentType) {
        var metadatas = reflector.annotations(componentType);
        var hostViewFactory = metadatas.find(isHostViewFactory);
        if (isBlank(hostViewFactory)) {
            throw new BaseException(`No precompiled component ${stringify(componentType)} found`);
        }
        return PromiseWrapper.resolve(new HostViewFactoryRef_(hostViewFactory));
    }
    clearCache() { }
};
Compiler_ = __decorate([
    Injectable(), 
    __metadata('design:paramtypes', [])
], Compiler_);
